### ISSUE TITLE
Builtin Time type

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4522]
This task is automatically imported from the old Task Issue Board and it was originally created by Hubert Plociniczak.
Original issue is [here](https://www.pivotaltracker.com/story/show/182973664).
<hr />

Similarly to Date [support](https://github.com/enso-org/enso/pull/3559) we would like to have a builtin Time type so that we can have proper polyglot support:
- with Enso Time
- with Javascript Time (as in miliseconds representation)
- with Javascript Time wrapper in Array
- with Java LocalDateTime allocated directly

We should potentially rename Time to DateTime (or ZonedDateTime?), to avoid confusion (it carries Date information as well).

For example
```
foreign js datetimeArr = """
    return [1, new Date('July 20, 69 20:17:40 GMT+00:00'), 7]

main =
    dtime = datetimeArr.at 1
    dtime.hour == 20
```
should return True (currently crashes).

## Comments:
**Hubert Plociniczak** reports a new **STANDUP** for the last Friday (2022-08-12):

**Progress:** Pattern matching on constants got merged. Spent most of the day exploring lazy visualization and exploring date/time-related support and drafting tickets for the future work. It should be finished by 2022-08-17.

**Next Day:** Next day I will be working on the [##182973664 task](https://www.pivotaltracker.com/story/show/182973664). Finish drafting tickets for date/time-related support, potentially pick up lazy visualization in parallel. (Enso Bot - Aug 15, 2022)
<hr />
Will also need to make Zone a builtin type to make it Zone-aware (Hubert Plociniczak - Aug 15, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-08-15):

**Progress:** Working on builtin DateTime type, good progress, should have some WIP PR tomorrow. It should be finished by 2022-08-17.

**Next Day:** Next day I will be working on the [##182973664 task](https://www.pivotaltracker.com/story/show/182973664). Continue work on DateTime, address review comments in Vector/polyglot array PR (Enso Bot - Aug 16, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-08-16):

**Progress:** Continued working on DateTime. Turned out it was easier to also address Time_Of_Day and Zone types as well, rather than moving/modifying existing tests. It should be finished by 2022-08-17.

**Next Day:** Next day I will be working on the [##182973664 task](https://www.pivotaltracker.com/story/show/182973664). Continue work on time/date polyglot support, address review comments in Vector/polyglot array PR (Enso Bot - Aug 17, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for today (2022-08-17):

**Progress:** Had a bit of fun with getting Javascript tests to accept the right timezone in tests. Once I got DateTime builtin type working and their tests passing.... I discovered that I broke Date polyglot support for JS. This should have been expected because JS `new Date(...)` is not converted to a LocalDate anymore but to ZonedDateTime (now we check `isDate`, `isTime`, `isTimezone` of the polyglot value). It worked before because the conversion was trivial with a single Date builtin type. Will have to adapt tests (only JS part, Java works as expected because it has LocalTime and LocalDate types). This will delay the delivery by a day or two. It should be finished by 2022-08-17.

**Next Day:** Next day I will be working on the [##182973664 task](https://www.pivotaltracker.com/story/show/182973664). (unfortunately) Continue work on time/date polyglot support, address review comments in Vector/polyglot array PR (Enso Bot - Aug 17, 2022)
<hr />
**Hubert Plociniczak** reports a new **🔴 DELAY** for yesterday (2022-08-18):

**Summary:** There is **2 days** delay in implementation of the [Builtin Time type (#182973664) task](https://www.pivotaltracker.com/story/show/182973664).
It **will cause 0 days delay** for the delivery of this weekly plan.

**Delay Cause:** Apart from Enso Date_Time also tackled Time_Of_Day and Zone, and the whole setup wasn't as trivial as it seemed initially. (Enso Bot - Aug 19, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-08-18):

**Progress:** All tests are now passing for new date/time types. Had to add more tweaks to data formatters for table, in a similar manner that was done for date. PR is ready for review. It should be finished by 2022-08-19.

**Next Day:** Next day I will be working on the [##182973664 task](https://www.pivotaltracker.com/story/show/182973664). Address PR review, pick up next item on my list (Arrays + lazy visualizations). (Enso Bot - Aug 19, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for the last Friday (2022-08-19):

**Progress:** More tests and fixes for reading/writing data from tables. Picked up Vector/polyglot array PR again, did some more testing. It should be finished by 2022-08-19.

**Next Day:** Next day I will be working on the [##182973664 task](https://www.pivotaltracker.com/story/show/182973664). Pick up next item on my list (Arrays or lazy visualizations). (Enso Bot - Aug 22, 2022)
<hr />
